src/perform_step/imex_perform_step.jl

```diff
@@ -0,0 +1,259 @@
+function initialize!(integrator, cache::CNABConstantCache)
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+@muladd function perform_step!(integrator, cache::CNABConstantCache, repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack uf,κ,tol = cache
+
+  if typeof(integrator.f) <: SplitFunction
+    f = integrator.f.f1
+    f2 = integrator.f.f2
+  else
+    f = integrator.f
+  end
+
+  # precalculations
+  κtol = κ*tol
+
+  dto2 = dt/2
+
+  # calculate W
+  uf.t = t
+  if typeof(uprev) <: AbstractArray
+    J = ForwardDiff.jacobian(uf,uprev)
+    W = I - dto2*J
+  else
+    J = ForwardDiff.derivative(uf,uprev)
+    W = 1 - dto2*J
+  end
+
+  # initial guess
+  zprev = 0
+  if typeof(integrator.f) <: SplitFunction
+    zprev = dt.*f(uprev, p, t)
+  else
+    zprev = dt*integrator.fsalfirst
+  end
+  z = zprev # Constant extrapolation
+
+  # initial step of Newton iteration
+  iter = 1
+  tstep = t + dt
+  tmp = uprev + dto2*integrator.fsalfirst
+
+  if typeof(integrator.f) <: SplitFunction
+    # This assumes the implicit part is cheaper than the explicit part
+    y0 = dt*integrator.fsalfirst - z
+    tmp += (-1/2)*y0
+  end
+
+  u = tmp + z/2
+  b = dt*f(u, p, tstep) - z
+  dz = W\b
+  ndz = integrator.opts.internalnorm(dz)
+  z = z + dz
+
+  η = max(cache.ηold,eps(eltype(integrator.opts.reltol)))^(0.8)
+  do_newton = integrator.success_iter == 0 || η*ndz > κtol
+
+  # Newton iteration
+  fail_convergence = false
+  while (do_newton || iter < integrator.alg.min_newton_iter) && iter < integrator.alg.max_newton_iter
+    iter += 1
+    u = tmp + z/2
+    b = dt*f(u, p, tstep) - z
+    dz = W\b
+    ndzprev = ndz
+    ndz = integrator.opts.internalnorm(dz)
+    θ = ndz/ndzprev
+    if θ > 1 || ndz*(θ^(integrator.alg.max_newton_iter - iter)/(1-θ)) > κtol
+      fail_convergence = true
+      break
+    end
+    η = θ/(1-θ)
+    do_newton = (η*ndz > κtol)
+    z = z + dz
+  end
+
+  if (iter >= integrator.alg.max_newton_iter && do_newton) || fail_convergence
+    integrator.force_stepfail = true
+    return
+  end
+
+  if typeof(integrator.f) <: SplitFunction
+    u = temp + z/2
+    y1 = dt*f2(u,p,t + dt)
```

Comment:
This shouldn't be using the new `u`, and it shouldn't be using `t+dt = t_{n+1}`

Reply:
@ChrisRackauckas in the simple trapezoid rule before the newton iteration.
`u = tmp + z/2` this step does the `g(tn)` right and the z we get after the newton iteration is the `g(tn+1)` 

In kencarp after the 2nd step the calculation is.
   `u = tmp + γ*z₂
    k2 = dt*f2(u,p,t + 2γ*dt)
    tmp = uprev + a31*z₁ + a32*z₂ + ea31*k1 + ea32*k2`

For my method I also went along similar lines 
The u = temp + z/2 
The 2nd line should I be using zprev instead of u right as that is the un
`y1 = dt*f2(zprev,p,t)`

Reply:
`z = dt*f(t+dt,u_{n+1})`. See line 72 to see what it's iterating. It will iterate until `b = dt*f(u, p, tstep) - z = 0`

Reply:
The reference for this way of writing it is https://www.sciencedirect.com/science/article/pii/0168927495001158